### PR TITLE
perf(spans): Prevent span extraction when quota is active

### PR DESCRIPTION
**Pull request by jjbayer**
Track spans inside a transaction item in `EnvelopeSummary::span_quantity`, and skip their extraction if the `span_indexed` category is rate limited.

This should reduce load on redis.

NOTE: With this PR, Relay will produce negative outcomes for spans in _any_ transaction, even for projects that do not have span extraction enabled (AM1). This should be fine, the product will simply ignore these outcomes.

### CODE REVIEW

File: relay-server/src/metrics_extraction/event.rs [hunk 1]
```diff
@@ -49,14 +49,12 @@ impl Extractable for Span {
...
     event: &mut Event,
-    spans_extracted: bool,
     config: CombinedMetricExtractionConfig<'_>,
     max_tag_value_size: usize,
     span_extraction_sample_rate: Option<f32>,
 ) -> Vec<Bucket> {
     let mut metrics = generic::extract_metrics(event, config);
-    // If spans were already extracted for an event, we rely on span processing to extract metrics.
-    if !spans_extracted && sample(span_extraction_sample_rate.unwrap_or(1.0)) {
+    if sample(span_extraction_sample_rate.unwrap_or(1.0)) {
```
Comment (jjbayer) [inline relay-server/src/metrics_extraction/event.rs:57]: We needed this check to avoid double extraction for transactions that were created from spans, feature that [has been removed](https://github.com/getsentry/relay/pull/3801).

File: relay-server/src/services/buffer/mod.rs [hunk 2]
```diff
@@ -220,11 +220,8 @@ impl EnvelopeBufferService {
         services: &Services,
         envelopes_tx_permit: Permit<'a, DequeuedEnvelope>,
     ) -> Result<Duration, EnvelopeBufferError> {
-        relay_log::trace!("EnvelopeBufferService: peeking the buffer");
```
Comment (jjbayer) [inline relay-server/src/services/buffer/mod.rs:223]: Changes in the file are unrelated, I got irritated by the amount of log lines I saw during integration testing.

File: relay-server/src/services/project.rs [hunk 3]
```diff
@@ -557,19 +555,9 @@ impl Project {
...
 
-        let check_nested_spans = state
-            .as_ref()
-            .is_some_and(|s| s.has_feature(Feature::ExtractSpansFromEvent));
-
-        // If we can extract spans from the event, we want to try and count the number of nested
-        // spans to correctly emit negative outcomes in case the transaction itself is dropped.
-        if check_nested_spans {
-            sync_spans_to_enforcement(&envelope, &mut enforcement);
-        }
```
Comment (jjbayer) [inline relay-server/src/services/project.rs:571]: This is now handled by the envelope summary.

File: relay-server/src/envelope.rs [hunk 4]
```diff
@@ -871,6 +872,27 @@ impl Item {
...
+        #[derive(Debug, Deserialize)]
+        struct PartialEvent {
+            spans: SeqCount,
+        }
+
+        if self.ty() != &ItemType::Transaction || self.spans_extracted() {
+            return 0;
+        }
+
+        serde_json::from_slice::<PartialEvent>(&self.payload()).map_or(0, |event| event.spans.0 + 1)
```
Comment (Dav1dde) [inline relay-server/src/envelope.rs:893]: Should we persist this in a header and/or memoize it?

File: relay-server/src/utils/rate_limits.rs [hunk 5]
```diff
@@ -458,6 +462,12 @@ impl Enforcement {
             return false;
         }
 
+        if item.ty() == &ItemType::Transaction && self.spans_indexed.is_active() {
+            // We cannot remove nested spans from the transaction, but we can prevent them
+            // from being extracted into standalone spans.
+            item.set_spans_extracted(true);
```
Comment (Dav1dde) [inline relay-server/src/utils/rate_limits.rs:468]: Seems quite hacky but I can't think of anything better.

**Review by Dav1dde — APPROVED** [review]

File: relay-server/src/envelope.rs (hunk 4 above)
Comment (jjbayer) [reply relay-server/src/envelope.rs:893]: We could if necessary, but that would mean we have to update it as well (e.g. trimming processor which removes spans), so I'm hesitant to add it now.

File: relay-server/src/utils/rate_limits.rs (hunk 5 above)
Comment (jjbayer) [reply relay-server/src/utils/rate_limits.rs:468]: We could rename the flag to `skip_span_extraction`. Would that help?
